Null pointer dereference in WebCore::ImageDocument::createDocumentStructure.

lericaa opened the pull request:
#### bac31e80659aa2caacbacfbba84a4053e322ed6f
<pre>
Null pointer dereference in WebCore::ImageDocument::createDocumentStructure.
<a href="https://bugs.webkit.org/show_bug.cgi?id=270283.">https://bugs.webkit.org/show_bug.cgi?id=270283.</a>
<a href="https://rdar.apple.com/122779661">rdar://122779661</a>.

Reviewed by Chris Dumez.

Adding null check to prevent the cases where local frame would be detached during createDocumentStructure.

* LayoutTests/fast/images/imageDocument-appendBytes-crash-expected.txt: Added.
* LayoutTests/fast/images/imageDocument-appendBytes-crash.html: Added.
* Source/WebCore/html/ImageDocument.cpp:
(WebCore::ImageDocument::updateDuringParsing):
(WebCore::ImageDocument::createDocumentStructure):

Canonical link: <a href="https://commits.webkit.org/275537@main">https://commits.webkit.org/275537@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f205caaef219cb0c2219bbb790fa2072840ce598

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/42117 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/21135 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44511 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/44704 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/38230 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/44424 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/24340 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/18464 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/34875 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42691 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/18061 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/36263 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/15807 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/15727 "Unexpected infrastructure issue, retrying build") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/37298 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/46137 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/38302 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37625 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/41521 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/16933 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/13919 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/40081 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/18552 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/36568 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9429 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/18612 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/18197 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/html/ImageDocument.cpp:
```diff
@@ -147,6 +147,9 @@ void ImageDocument::updateDuringParsing()
     if (!m_imageElement)
         createDocumentStructure();
 
+    if (RefPtr frame = document().frame(); !frame)
```
You don’t need to ref the frame just to null check it.

`if (!document().frame())`

cdumez commented on Source/WebCore/html/ImageDocument.cpp:
```diff
@@ -147,6 +147,9 @@ void ImageDocument::updateDuringParsing()
     if (!m_imageElement)
         createDocumentStructure();
 
+    if (!document().frame())
```
I believe `document()` returns `this` here given this `this` is an ImageDocument (which is a Document). Therefore, we can probably use `if (!frame())`

lericaa replied:
yes, i feel the same and was confused by the usage of `this` in other functions from this ImageDocument. addressed in latest patch.

lericaa commented on Source/WebCore/html/ImageDocument.cpp:
```diff
@@ -249,8 +253,8 @@ void ImageDocument::createDocumentStructure()
-    if (auto* cachedImage = imageElement->cachedImage())
-        cachedImage->setResponse(loader()->response());
+    if (auto* cachedImage = imageElement->cachedImage(); documentLoader && cachedImage)
+        cachedImage->setResponse(documentLoader->response());
```
Add the null check here also as it is referenced.

lericaa committed:
Null pointer dereference in WebCore::ImageDocument::createDocumentStructure.
https://bugs.webkit.org/show_bug.cgi?id=270283.
rdar://122779661.

Reviewed by Chris Dumez.

Adding null check to prevent the cases where local frame would be detached during createDocumentStructure.

* LayoutTests/fast/images/imageDocument-appendBytes-crash-expected.txt: Added.
* LayoutTests/fast/images/imageDocument-appendBytes-crash.html: Added.
* Source/WebCore/html/ImageDocument.cpp:
(WebCore::ImageDocument::updateDuringParsing):
(WebCore::ImageDocument::createDocumentStructure):

Canonical link: https://commits.webkit.org/275537@main
